### PR TITLE
http: Convert HttpInputStreamImpl to coroutines

### PR DESCRIPTION
**Pull request by harrishancock**
This required a slight adjustment to a test, due to a change in the eagerness of a promise chain.

### CODE REVIEW

**Comment by harrishancock** [comment]
I recommend hiding whitespace changes to review.

File: c++/src/kj/compat/http-test.c++ [hunk 1]
```diff
@@ -1175,6 +1175,12 @@ KJ_TEST("HttpClient chunked body pump from fixed length stream") {
     kj::StringPtr responseText = "HTTP/1.1 204 No Content\r\n\r\n";
     pipe.ends[1]->write(responseText.begin(), responseText.size()).wait(waitScope);
     auto response = req.response.wait(waitScope);
+
+    // Wait even harder, because the coroutine conversion implicitly eagerly-evaluates stuff, which
+    // I guess makes the response complete before the damn terminal chunk now.
+    //
+    // TODO(2.0): req.body->end() should fix this, right?
+    evalLater([]() {}).wait(waitScope);
```
Comment (kentonv) [inline c++/src/kj/compat/http-test.c++:1183]: Maybe `serverPromise.poll(waitScope)` instead? Heck, you could remove the whole response part above and replace it with that.

It is strange that the code now processes the entire response without stopping to finish the request.

**Review by kentonv — APPROVED** [review]

**Comment by harrishancock** [comment]
Benchmark results with `http-over-capnp-perf-test -b1000000:
An old v2 branch HEAD:
```
[ TEST ] http-over-capnp-perf-test.c++:362: Benchmark KJ HTTP full protocol
capnp/compat/http-over-capnp-perf-test.c++:75: mallocCount = 55200054; mallocBytes = 42907823791
```
This PR:
```
[ TEST ] http-over-capnp-perf-test.c++:362: Benchmark KJ HTTP full protocol
capnp/compat/http-over-capnp-perf-test.c++:75: mallocCount = 66200052; mallocBytes = 56363822127
```

**Comment by harrishancock** [comment]
Merging, as the performance regression seems specific to that one test -- in the Workers runtime there is no performance difference at all.

**Pull request merged**